horqrux/shots.py

```diff
@@ -0,0 +1,54 @@
+from __future__ import annotations
+
+from functools import reduce
+from typing import Any
+
+import jax
+import jax.numpy as jnp
+from jax import Array
+from jax.experimental import checkify
+
+from horqrux.apply import apply_gate
+from horqrux.primitive import GateSequence, Primitive
+from horqrux.utils import none_like
+
+
+def observable_to_matrix(observable: Primitive, n_qubits: int) -> Array:
+    """For finite shot sampling we need to calculate the eigenvalues/vectors of
+    an observable. This helper function takes an observable and system size
+    (n_qubits) and returns the overall action of the observable on the whole
+    system.
+
+    LIMITATION: currently only works for observables which are not controlled.
+    """
+    checkify.check(
+        observable.control == observable.parse_idx(none_like(observable.target)),
+        "Controlled gates cannot be promoted from observables to operations on the whole state vector",
+    )
+    unitary = observable.unitary()
+    target = observable.target[0][0]
+    identity = jnp.eye(2, dtype=unitary.dtype)
+    ops = [identity for _ in range(n_qubits)]
+    ops[target] = unitary
+    return reduce(lambda x, y: jnp.kron(x, y), ops[1:], ops[0])
+
+
+def finite_shots_fwd(
+    state: Array,
+    gates: GateSequence,
+    observable: Primitive,
+    values: dict[str, float],
+    n_shots: int = 100,
+    key: Any = jax.random.PRNGKey(0),
+) -> Array:
+    """
+    Run 'state' through a sequence of 'gates' given parameters 'values'
+    and compute the expectation given an observable.
+    """
+    state = apply_gate(state, gates, values)
+    n_qubits = len(state.shape)
+    mat_obs = observable_to_matrix(observable, n_qubits)
```

Comment:
This promotion is to easily get the eigenvectors of the hamiltonian as it acts on the entire statevector, and not just on the target qubit.

Reply:
I'm not sure if there's an elegant way to get the eigenvectors without having to promote the hamiltonian in such a manner.

Reply:
Hmm I suspect you can, isn't it just the eigenvector matrix of the small hamiltonian kronned with the identity matrix or something? At least it seems to me like you can get a valid eigendecomposition that way:

$$
H_\text{big} = H_\text{small} \otimes I = \left(\sum_\lambda \lambda|\lambda\rangle\langle\lambda|\right)\otimes I
$$

where each eigenvalue of $H_\text{small}$ will show up repeated by the size of the identity. Check Theorem 13.12 [here](https://archive.siam.org/books/textbooks/OT91sample.pdf).

Reply:
Nice! I've raised an issue to investigate this. It's _possible_ that the XLA compiler will be able to optimise this away when compiling: `eigh` will always be called with the same arguments in most applications, so it may well cache the result.